### PR TITLE
Speed up transaction account range query performance by improving db schema

### LINKED ISSUE #288: /transactions query is slow on mainnet with account.id range
This query takes over 5 seconds on mainnet: /transactions?account.id=gt:18000&account.id=lt:19900

**Detailed Description**
transactions query with account id range takes long.

**Actual Behavior**
1. Query "/transactions?account.id=gt:18000&account.id=lt:19900" on mainnet

**Expected Behavior**
It should return in under 1 second

**Environment:**
- Version: [e.g. 0.2.0]

### PR DESCRIPTION
**Pull request by Nana-EC**
**Detailed description**:
Transaction rest-api queries of the account range type are slow and take too long. Averaging about 8.2 seconds for a query of the following kind 
/transactions?account.id=gt:0&account.id=lt:1900

Looking into it the cause is an inner sql join on the t_cryptotransferlists and the t_entities table to get the account realm and num.
Per Mike's suggestion this happens often enough that we may duplicate some data from t_entities in t_cryptotransferlists to improve the speed.

This PR increase the speed by doing so and is able to cut the time down by ~ 75% i.e from 8.2 secs to about 2.2 seconds

Changes inclue
-  Adding new flywaydb migration version to add account_realm_num and account_num as well as an index to t_entitites
- Updating CryptoTransfer domain with new fields
- Updating RecordFileLooger with enums to map to new columns and updating sql table insertions with new column information
- Updating java UT's to account for change
- Updating CryptoTransferRepository with customized search method based off of new columns

**Which issue(s) this PR fixes**:
Fixes #288 

**Special notes for your reviewer**:
Note initially we thought account id could be removed but this must remain as the query value refers to the account id.

**Checklist**
- [ ] Documentation added
- [x] Tests updated

### CODE REVIEW

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql [hunk 1]
```diff
@@ -0,0 +1,15 @@
+---
+--- Replace t_cryptotransferlists.account_id with account_realm_num, and account_num
+---
+alter table t_cryptotransferlists
+    add column if not exists account_realm_num bigint null;
```
Comment (steven-sheehy) [inline src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: I think the realm applies to not just the account but the cryptotransfer as well, so we can just call it `realm`. I think it more of as a shard key that applies to nearly all tables instead of just something attached to the account.

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+alter table t_cryptotransferlists
+    add column if not exists account_num bigint null;
+
+with dupe_entities as (
+        select id, entity_realm, entity_num FROM t_entities
+)
+update t_cryptotransferlists c SET account_realm_num = de.entity_realm, account_num = de.entity_num FROM dupe_entities de WHERE c.account_id = de.id;
```
Comment (steven-sheehy) [inline src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: There's a mix of case in this file. Please convert everything to lowercase

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 1 above)
Comment (steven-sheehy) [inline src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: It should probably be of type `entity_realm_num`

File: src/main/java/com/hedera/mirror/repository/CryptoTransferRepository.java [hunk 3]
```diff
@@ -27,5 +27,6 @@
 import com.hedera.mirror.domain.CryptoTransfer;
 
 public interface CryptoTransferRepository extends CrudRepository<CryptoTransfer, Long> {
-    Optional<CryptoTransfer> findByConsensusTimestampAndAccountId(Long consensusTimestamp, Long accountId);
+	Optional<CryptoTransfer> findByConsensusTimestampAndAccountId(Long consensusTimestamp, Long accountId);
+    Optional<CryptoTransfer> findByConsensusTimestampAndAccountNum(Long consensusTimestamp, Long accountNum);
```
Comment (steven-sheehy) [inline src/main/java/com/hedera/mirror/repository/CryptoTransferRepository.java]: New repository methods need a corresponding test.

File: src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java [hunk 4]
```diff
@@ -540,7 +541,7 @@ public static void storeRecord(Transaction transaction, TransactionRecord txReco
 
         if ((txRecord.hasTransferList()) && parserProperties.isPersistCryptoTransferAmounts()) {
             if (body.hasCryptoCreateAccount() && isSuccessful(txRecord)) {
-                insertCryptoCreateTransferList(consensusNs, txRecord, body, createdAccountId, fkPayerAccountId);
+                insertCryptoCreateTransferList(consensusNs, txRecord, body, txRecord.getReceipt().getAccountID(), fkPayerAccountId);
```
Comment (steven-sheehy) [inline src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java]: Why use `txRecord.getReceipt().getAccountID()` instead of `createdAccountId` stored earlier? If this is the only place `createdAccountId` is/was used then I'd say delete that variable and keep this how you have it.

File: src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java [hunk 5]
```diff
@@ -677,18 +682,22 @@ private static void insertCryptoCreateTransferList(final long consensusTimestamp
...
+            var accountId = aa.getAccountID();
+			long account = accountId.getAccountNum();
 
             sqlInsertTransferList.setLong(F_TRANSFERLIST.ACCOUNT_ID.ordinal(),
                     entities.createOrGetEntity(aa.getAccountID()));
             sqlInsertTransferList.setLong(F_TRANSFERLIST.AMOUNT.ordinal(), amount);
+            sqlInsertTransferList.setLong(F_TRANSFERLIST.ACCOUNT_REALM_NUM.ordinal(),
+                    accountId.getRealmNum());
+            sqlInsertTransferList.setLong(F_TRANSFERLIST.ACCOUNT_NUM.ordinal(),
+                    accountId.getAccountNum());
```
Comment (steven-sheehy) [inline src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java]: Use `account`

File: src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java [hunk 6]
```diff
@@ -698,13 +707,23 @@ private static void insertCryptoCreateTransferList(final long consensusTimestamp
 
         if (addInitialBalance) {
             sqlInsertTransferList.setLong(F_TRANSFERLIST.CONSENSUS_TIMESTAMP.ordinal(), consensusTimestamp);
-            sqlInsertTransferList.setLong(F_TRANSFERLIST.ACCOUNT_ID.ordinal(), payerAccountId);
+            sqlInsertTransferList.setLong(F_TRANSFERLIST.ACCOUNT_ID.ordinal(),
```
Comment (steven-sheehy) [inline src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java]: Do we really need to store account id, account num and account realm in crypto transfers? Can't we remove the account ID now?

File: src/test/java/com/hedera/recordFileLogger/RecordFileLoggerCryptoTest.java [hunk 7]
```diff
@@ -84,945 +84,1141 @@
...
-                , () -> assertNull(dbNewAccountEntity.getExpiryTimeNs())
+		assertTrue(RecordFileLogger.start());
+		assertEquals(INIT_RESULT.OK, RecordFileLogger.initFile("TestFile"));
+	}
+
+	@AfterEach
+	void after() {
+		RecordFileLogger.finish();
+	}
+
```
Comment (steven-sheehy) [inline src/test/java/com/hedera/recordFileLogger/RecordFileLoggerCryptoTest.java]: Why is there so much whitespace change in this file? Make sure you have the save actions plugin installed.

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+create index if not exists idx_t_cryptotransferlists_unq ON t_cryptotransferlists (account_realm_num, account_num, consensus_timestamp desc);
```
[the remark this answers is not in the document]
Comment (mike-burrage-hedera) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: actually - do those drop indexes first in the file.

File: src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java (hunk 4 above)
Comment (Nana-EC) [reply src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java]: Yeah java compiler complains because there's a potential flow where this variable is uninitialized. I can make it consistent

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 1 above)
Comment (Nana-EC) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: I think Mike and I agreed on leaving it as is.

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 2 above)
Comment (Nana-EC) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: missed some of those, fixed

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 1 above)
Comment (Nana-EC) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: Fixed, didn't know about this type

Comment (Nana-EC) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: Notably t_entities uses bitint

File: src/main/java/com/hedera/mirror/repository/CryptoTransferRepository.java (hunk 3 above)
Comment (Nana-EC) [reply src/main/java/com/hedera/mirror/repository/CryptoTransferRepository.java]: I have one in the same location that findByConsensusTimestampAndAccountId() does in the AbstractRecordLoggerTests class. 
Are you suggesting another location?

Comment (Nana-EC) [reply src/main/java/com/hedera/mirror/repository/CryptoTransferRepository.java]: Also RecordFileLoggerCryptoTest uses tests for it.

Comment (steven-sheehy) [reply src/main/java/com/hedera/mirror/repository/CryptoTransferRepository.java]: No, you need add a test in `CryptoTransferRepositoryTest` (create if doesn't exist).

File: src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java (hunk 5 above)
Comment (Nana-EC) [reply src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java]: Done

File: src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java (hunk 6 above)
Comment (Nana-EC) [reply src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java]: Will remove account_id given recent offline discussion on the sql inner query improvement.

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: Done

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql [hunk 9]
```diff
@@ -0,0 +1,15 @@
+---
+--- Replace t_cryptotransferlists.account_id with account_realm_num, and account_num
+---
+alter table t_cryptotransferlists
+    add column if not exists account_realm_num bigint null;
+
+alter table t_cryptotransferlists
+    add column if not exists account_num bigint null;
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: Done

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+
+with dupe_entities as (
+        select id, entity_realm, entity_num FROM t_entities
+)
+update t_cryptotransferlists c SET account_realm_num = de.entity_realm, account_num = de.entity_num FROM dupe_entities de WHERE c.account_id = de.id;
+
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: Added

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 1 above)
Comment (apeksharma) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: i had the same comment.
1. 'realm' and 'num' are separate. So let's not use both together in one name. Suggestions: 'realmId' (https://docs.hedera.com/docs/accounts#section-account-id)
2. The 'shard.realm.num' getting the hbars can be of 'account' type or 'contract' type. So let's put the type of entity in the column name.
'realm' or 'realmId' is probably the right name here.

File: rest-api/transactions.js [hunk 11]
```diff
@@ -292,9 +290,8 @@ const getOneTransaction = function(req, res) {
     '   , enode.entity_realm as node_realm\n' +
```
Comment (apeksharma) [inline rest-api/transactions.js]: same, maybe we are able to delete entity_shard and node_shard here too?

File: src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java (hunk 4 above)
Comment (apeksharma) [reply src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java]: If it's possible, let's move the call to `insertCryptoCreateTransferList(...)` after line 398?
```
if (txRecord.hasTransferList() && parserProperties.isPersistCryptoTransferAmounts()) {
    insertCryptoCreateTransferList(consensusNs, txRecord, body, createdAccountId, fkPayerAccountId);
}
```

File: src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java [hunk 12]
```diff
@@ -677,18 +682,22 @@ private static void insertCryptoCreateTransferList(final long consensusTimestamp
...
-            payerAccountNum = body.getTransactionID().getAccountID().getAccountNum();
         }
         TransferList transferList = txRecord.getTransferList();
         for (int i = 0; i < transferList.getAccountAmountsCount(); ++i) {
             sqlInsertTransferList.setLong(F_TRANSFERLIST.CONSENSUS_TIMESTAMP.ordinal(), consensusTimestamp);
             var aa = transferList.getAccountAmounts(i);
             long amount = aa.getAmount();
-            long account = aa.getAccountID().getAccountNum();
+            var accountId = aa.getAccountID();
+			long account = accountId.getAccountNum();
```
Comment (apeksharma) [inline src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java]: nit: indentation

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 2 above)
Comment (apeksharma) [inline src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: How much time should it take to update 100M+ rows in the table?
Is there a possibility of any timeouts - connection/migration/etc?
Basically, should we worry about a case like https://stackoverflow.com/a/25187636?

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (apeksharma) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: We should compare query plans first and then decide the order.
The CTE based population of new columns above is already not a good way, it won't be able to use the `id` index in t_entities (https://medium.com/@hakibenita/be-careful-with-cte-in-postgresql-fca5e24d2119).
Then if we also drop the indexes in t_cyptotransferlists, I think it'll be disastrous.

Let's first compare query plans for:
1. the update expression above (using CTE)
2. simple update join

**Comment by apeksharma** [comment]
@steven-sheehy  @mike-burrage-hedera  how many known external mirror nodes are there?
What's their upgrade story?

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: Was that first index a replacement or in addition to the one I added in here already?

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 1 above)
Comment (mike-burrage-hedera) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: https://github.com/hashgraph/hedera-protobuf/blob/master/src/main/proto/BasicTypes.proto

In the protobuf they are called realmNum and shardNum and accountNum/fileNum etc.

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (mike-burrage-hedera) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: First was in addition as the outer query is by consensus_ts.

[the remark this answers is not in the document]
Comment (mike-burrage-hedera) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: As for comparing the query plans before deciding order (of I assume the drop index) - why?
This is an upgrade script where we're modifying all the rows in the table and dropping indexes. Why wouldn't we drop the index first?

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 1 above)
Comment (mike-burrage-hedera) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: > @steven-sheehy @mike-burrage-hedera how many known external mirror nodes are there?
> What's their upgrade story?

Per @gregscullard, there are 5 or 6 processes that are retrieving streams from the buckets. I think far fewer are using our database schema.

**Comment by mike-burrage-hedera** [comment]
> @steven-sheehy @mike-burrage-hedera how many known external mirror nodes are there?
> What's their upgrade story?

Per @gregscullard, there are 5 or 6 processes that are retrieving streams from the buckets. I think far fewer are using our database schema.

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (mike-burrage-hedera) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: Another other option is to:

- make a fresh table `t_cryptotransferlists_new`
- insert into it, instead of updated the existing table
- then drop the old and rename the _new table
- apply indexes

This would be the "cleanest" way to clean up the table and would likely result in the smallest end result (in terms of data storage for the table) and no need for vacuuming, but may also take too long. Probably negligible impact on insert/query performance, but less data storage used.

All these optimizations around the table/data upgrade can probably be left until the last step being worked on.

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 1 above)
Comment (apeksharma) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: 'realm' and 'num' go together in proto, fair enough.

> 'shard.realm.num' can be BOTH an account and a contract.
That's what i said in point 2. It can be of either type.
So let's not prefix 'realm' by 'account'. should be just 'realmNum'/'realm'/'realmId'.

Unlike in proto, which can have different message type, we don't have the luxury of being able to define different columns for accounts/files/contract.
I really can't think of a better name for `account_num` which is agnostic of entity type (account). Can you suggest one @steven-sheehy ?

Comment (steven-sheehy) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: @apeksharma We call the account/file/contract/topic an `entity`, so it would be `entity_num`.

I'm okay with `realm_num`, just not with `account_realm_num`

File: rest-api/transactions.js [hunk 13]
```diff
@@ -146,7 +145,6 @@ const getTransactionsInnerQuery = function(accountQuery, tsQuery, resultTypeQuer
     '       from t_cryptotransferlists ctl\n' +
     '       join t_transactions t on t.consensus_ns = ctl.consensus_timestamp\n' +
     '       join t_transaction_results tr on t.fk_result_id = tr.id\n' +
-    '       join t_entities eaccount on eaccount.id = ctl.account_id\n' +
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply rest-api/transactions.js]: Done

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (Nana-EC) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: Made adjustments. Will leave the larger items for a separate db improvement PR as Mike hinted. Unless others feel strongly

File: rest-api/__tests__/integration.test.js [hunk 14]
```diff
@@ -199,8 +199,8 @@ const addTransaction = async function(
   for (var i = 0; i < transfers.length; ++i) {
     let xfer = transfers[i];
     await sqlConnection.query(
-      'insert into t_cryptotransferlists (consensus_timestamp, account_id, amount) values ($1, $2, $3);',
-      [consensusTimestamp, accountEntityIds[xfer[0]], xfer[1]]
+      'insert into t_cryptotransferlists (consensus_timestamp, account_id, amount, account_realm_num, account_num) values ($1, $2, $3, $4, $5);',
+      [consensusTimestamp, accountEntityIds[xfer[0]], xfer[1], realm, accountEntityIds[xfer[0]]]
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply rest-api/__tests__/integration.test.js]: Yep your right. Due to the accountEntityIds population logic it ends up being the same.
Made the  change cause it's cleaner

File: rest-api/transactions.js [hunk 15]
```diff
@@ -102,9 +103,8 @@ const getTransactionsOuterQuery = function(innerQuery, order) {
     '   , enode.entity_realm as node_realm\n' +
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply rest-api/transactions.js]: By definition they could be different, but by current design they won't be. However I don't think it'd be wise to mix them up now

File: rest-api/transactions.js (hunk 11 above)
Comment (Nana-EC) [reply rest-api/transactions.js]: Different entity, same thinking as above comment

File: src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java (hunk 12 above)
Comment (Nana-EC) [reply src/main/java/com/hedera/recordFileLogger/RecordFileLogger.java]: Fixed

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 1 above)
Comment (apeksharma) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: +1
`realm_num`
`entity_num`

More generally, going forward let's stick to these three naming convention in mirror node code (case adjusted to language)  : `shardNum` `realmNum` `entityNum`
That okay with you guys @mike-burrage-hedera  @Nana-EC ?

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (apeksharma) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: am not sql jedi, so my concern below may be wrong.
But I have serious concern about below sql 
```
with dupe_entities as (
        select id, entity_realm, entity_num from t_entities
)
update t_cryptotransferlists c set account_realm_num = de.entity_realm, account_num = de.entity_num from dupe_entities de where c.account_id = de.id;
```
since:
- it can't do full scan on t_ctl and do lookups on t_entities as index on `id` won't be available due to CTE
- neither t_entities, not t_ctl is small enough to be keep in memory and do full scan on other (unlike in steven's PR where `t_transaction_types` and `t_transaction_results` are very small and can be kept in memory)
- it can't do full scan on t_entities and do fast lookups on t_ctl since we have dropped t_ctl's indexes in beginning of this migration script

It's very likely that I may be wrong since my sql is not strong. But unless someone can explain why above query is okay, let's do an `EXPLAIN` on dev cluster and see the cost.
I just want to avoid the case where, we do an iteration on 60000 rows resulting from that CTE when updating every row in t_ctl.

[the remark this answers is not in the document]
Comment (apeksharma) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: So, i don't really worry about cleaner ways or _little more_ better ways, i just want to avoid worst case scenario - 100M ctl rows * 60-70k CTE rows.

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 2 above)
Comment (Nana-EC) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: So I did a query plan comparison and the processing time between this option and an update join is similar. The join was faster by only 2 secs.
I can share the comparison files offline if you'd like.

In this case are you recommending we change the timeout setting?

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 1 above)
Comment (mike-burrage-hedera) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: The reason I suggested account_realm_num is that it is the realm of the account to which is having its balance changed. The reason for my other comment was to correct "'realm' and 'num' are separate". Choose whichever name you like.

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (mike-burrage-hedera) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: If the table update is under 5 minutes on a production-level number of rows ... ship it.

[the remark this answers is not in the document]
Comment (apeksharma) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: Thanks @Nana-EC  for posting the numbers.
Posting query plans collected by Nana below for posterity:
```
QUERY PLAN
Update on t_cryptotransferlists c  (cost=36.83..381909.10 rows=6099398 width=88) (actual time=113885.587..113885.593 rows=0 loops=1)
  CTE dupe_entities
    ->  Seq Scan on t_entities  (cost=0.00..20.90 rows=490 width=24) (actual time=1.173..7.140 rows=526 loops=1)
  ->  Hash Join  (cost=15.93..381888.20 rows=6099398 width=88) (actual time=26.690..42770.531 rows=2075330 loops=1)
        Hash Cond: (c.account_id = de.id)
        ->  Seq Scan on t_cryptotransferlists c  (cost=0.00..101175.50 rows=2489550 width=30) (actual time=5.346..14746.865 rows=2075330 loops=1)
        ->  Hash  (cost=9.80..9.80 rows=490 width=72) (actual time=21.266..21.272 rows=526 loops=1)
              Buckets: 1024  Batches: 1  Memory Usage: 62kB
              ->  CTE Scan on dupe_entities de  (cost=0.00..9.80 rows=490 width=72) (actual time=1.249..16.546 rows=526 loops=1)
Planning time: 38.971 ms
Execution time: 113887.291 ms
```
```
QUERY PLAN
Update on t_cryptotransferlists  (cost=27.02..159190.84 rows=2986873 width=46) (actual time=115935.372..115935.378 rows=0 loops=1)
  ->  Hash Join  (cost=27.02..159190.84 rows=2986873 width=46) (actual time=19.169..43731.392 rows=2075330 loops=1)
        Hash Cond: (t_cryptotransferlists.account_id = t_entities.id)
        ->  Seq Scan on t_cryptotransferlists  (cost=0.00..121386.73 rows=2986873 width=30) (actual time=3.279..15363.307 rows=2075330 loops=1)
        ->  Hash  (cost=20.90..20.90 rows=490 width=30) (actual time=15.798..15.804 rows=526 loops=1)
              Buckets: 1024  Batches: 1  Memory Usage: 41kB
              ->  Seq Scan on t_entities  (cost=0.00..20.90 rows=490 width=30) (actual time=0.012..12.212 rows=526 loops=1)
Planning time: 19.377 ms
Execution time: 115936.698 ms
```

Few points:
Both plans above satisfy me since:
1. SQL optimizer decides to keep relevant t_entities data in memory (CTE result in first case, and some kind of pruned results in second case) and do a hash join on it.
2. In both the plans, keeping hash on CTE/t_entities is 62kb/41kb. That's for ~500 rows. In prod, we have ~18k rows. So it'll still be fine. (edited) 

However, the bigger concern now is - execution time.
It took about 100sec to update about 2mil rows in t_cryptotransferflists. In prod mainnet, we have 120 million rows.  We can't bring down prod for 1hr 45min.
So let's gets this PR committed and create a separate ticket for figuring out right way to migrate in next release.

File: src/test/java/com/hedera/mirror/parser/record/RecordFileLoggerCryptoTest.java [hunk 16]
```diff
@@ -87,776 +87,795 @@
     void before() throws Exception {
         parserProperties.setPersistClaims(true);
```
Comment (apeksharma) [inline src/test/java/com/hedera/mirror/parser/record/RecordFileLoggerCryptoTest.java]: This files seems to be undoing some of the whitespace changes that were done recently. It's also making some of the lines >120 chars which were fixed in the previous change.
your editor might need correction. Please sync up with @steven-sheehy, he might be able to help.

File: rest-api/transactions.js [hunk 17]
```diff
@@ -48,11 +49,11 @@ const createTransferLists = function(rows, arr) {
       transactions[row.consensus_ns]['name'] = row['name'];
       transactions[row.consensus_ns]['max_fee'] = utils.getNullableNumber(row['max_fee']);
       transactions[row.consensus_ns]['valid_duration_seconds'] = utils.getNullableNumber(row['valid_duration_seconds']);
-      transactions[row.consensus_ns]['node'] = row.node_shard + '.' + row.node_realm + '.' + row.node_num;
```
Comment (steven-sheehy) [inline rest-api/transactions.js]: I think node_shard can now be removed from the select since you removed its use.

File: src/main/java/com/hedera/mirror/parser/record/RecordFileLogger.java [hunk 18]
```diff
@@ -244,9 +247,9 @@ public static void storeRecord(Transaction transaction, TransactionRecord txReco
         sqlInsertTransaction.setLong(F_TRANSACTION.FK_REC_FILE_ID.ordinal(), fileId);
         sqlInsertTransaction.setLong(F_TRANSACTION.VALID_DURATION_SECONDS.ordinal(), validDurationSeconds);
 
-        long fkPayerAccountId = entities.createOrGetEntity(transactionID.getAccountID());
+        AccountID fkPayerAccountId = transactionID.getAccountID();
```
Comment (steven-sheehy) [inline src/main/java/com/hedera/mirror/parser/record/RecordFileLogger.java]: nit: Name should be payerAccountId since it's no longer the long foreign key.

File: src/test/java/com/hedera/mirror/parser/record/RecordFileLoggerCryptoTest.java (hunk 16 above)
Comment (Nana-EC) [reply src/test/java/com/hedera/mirror/parser/record/RecordFileLoggerCryptoTest.java]: Thanks enabled the plugin

File: src/test/java/com/hedera/mirror/repository/CryptoTransferRepositoryTest.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+import com.hedera.mirror.domain.RecordFile;
+import com.hedera.mirror.domain.Transaction;
+
+import org.junit.jupiter.api.Test;
+
+import com.hedera.mirror.domain.CryptoTransfer;
+
+import org.springframework.test.context.jdbc.Sql;
+
+@Sql(executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD, scripts = "classpath:db/scripts/cleanup.sql")
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply src/test/java/com/hedera/mirror/repository/CryptoTransferRepositoryTest.java]: Removed

File: src/test/java/com/hedera/mirror/repository/CryptoTransferRepositoryTest.java [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+public class CryptoTransferRepositoryTest  extends AbstractRepositoryTest {
+    @Test
+    void insert() {
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply src/test/java/com/hedera/mirror/repository/CryptoTransferRepositoryTest.java]: Changed

File: rest-api/transactions.js (hunk 17 above)
Comment (Nana-EC) [reply rest-api/transactions.js]: Done

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 1 above)
Comment (apeksharma) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: > The reason for my other comment was to correct "'realm' and 'num' are separate"
 👍 yes, that point was received.

tl;dr:
let's change column names to be `realm_num` and `entity_num` since the values in these column will not be of just 'accounts' type, but can also be of 'crypto' type.

File: src/main/java/com/hedera/mirror/parser/record/RecordFileLogger.java (hunk 18 above)
Comment (Nana-EC) [reply src/main/java/com/hedera/mirror/parser/record/RecordFileLogger.java]: Changed

File: src/test/java/com/hedera/recordFileLogger/RecordFileLoggerCryptoTest.java (hunk 7 above)
Comment (Nana-EC) [reply src/test/java/com/hedera/recordFileLogger/RecordFileLoggerCryptoTest.java]: Done. However mine was doing some funky things and removing a lot of the final classifiers to variables, which I hadn't seen in your PR's

File: src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql (hunk 1 above)
Comment (Nana-EC) [reply src/main/resources/db/migration/V1.14__cryptotransferlist_realm_num.sql]: Went with realm_num and entity_num. Thanks
Originally it sounded like a simple naming preference, but after you guys noted the protobuf relationship I'm convinced.

File: src/main/java/com/hedera/mirror/parser/record/RecordFileLogger.java (hunk 18 above)
Comment (steven-sheehy) [reply src/main/java/com/hedera/mirror/parser/record/RecordFileLogger.java]: Looks like you missed it in the commit, but it's minor so we can ignore
